journal/src/test/scala/com/evolutiongaming/kafka/journal/ActionHeaderJsonSpec.scala

```diff
@@ -8,75 +8,82 @@ import play.api.libs.json.*
 
-class ActionHeaderJsonSpec extends AnyFunSuite with Matchers {
+class ActionHeaderJsonSpec extends AnyFunSuite with Matchers with SerdeTesting {
 
-  val origins = List(Origin("origin").some, none)
-  val metadata = List(
+  // For some reason, ActionHeader Format is asymmetric:
+  // it has Writes for ActionHeader but Reads only for Option[ActionHeader].
+  // For this test, we can define Reads[ActionHeader] through the Option impl.
+  private implicit val actionHeaderReads: Reads[ActionHeader] = (json: JsValue) => {
+    ActionHeader.formatOptActionHeader.reads(json).map(_.get)
+  }
+  private implicit val actionHeaderFromBytes: FromBytes[Try, ActionHeader] = FromBytes.fromReads[Try, ActionHeader]
+
+  private val originExamples = Vector(Origin("origin").some, none)
+  private val metadataExamplesCurrentFormat = Vector(
     ("metadata", HeaderMetadata(Json.obj(("key", "value")).some)),
     ("none", HeaderMetadata.empty),
+  )
+  private val metadataExamplesLegacyFormat = Vector(
     ("legacy", HeaderMetadata.empty),
   )
+  private val payloadTypeExamples = Vector(PayloadType.Binary, PayloadType.Json)
 
-  val payloadTypes = List(PayloadType.Binary, PayloadType.Json)
+  private def appendExample(
+    origin: Option[Origin],
+    payloadType: PayloadType.BinaryOrJson,
+    metadata: HeaderMetadata,
+  ): ActionHeader = {
+    ActionHeader.Append(
+      range = SeqRange.unsafe(1, 5),
+      origin = origin,
+      version = none,
+      payloadType = payloadType,
+      metadata = metadata,
+    )
+  }
 
-  for {
-    origin <- origins
-  } {
+  originExamples.foreach { origin =>
     val originStr = origin.fold("None")(_.toString)
-    for {
-      payloadType <- payloadTypes
-      (metadataStr, metadata) <- metadata
-    } {
-      test(s"Append format, origin: $origin, payloadType: $payloadType, metadata: $metadataStr") {
-        val range = SeqRange.unsafe(1, 5)
-        val header =
-          ActionHeader.Append(
-            range = range,
-            origin = origin,
-            version = none,
-            payloadType = payloadType,
-            metadata = metadata,
-          )
-        verify(header, s"Append-$originStr-$payloadType-$metadataStr")
+
+    payloadTypeExamples.foreach { payloadType =>
+      metadataExamplesCurrentFormat.foreach { case (metadataStr, metadata) =>
+        test(s"Append current format, origin: $origin, payloadType: $payloadType, metadata: $metadataStr") {
+          val header = appendExample(origin, payloadType, metadata)
+          verifyEncodeDecodeExample(header, s"Append-$originStr-$payloadType-$metadataStr.json")
+        }
+      }
+
+      metadataExamplesLegacyFormat.foreach { case (metadataStr, metadata) =>
+        test(s"Append legacy format, origin: $origin, payloadType: $payloadType, metadata: $metadataStr") {
+          val header = appendExample(origin, payloadType, metadata)
+          verifyDecodeExample(header, s"Append-$originStr-$payloadType-$metadataStr.json")
```

Comment:
we have only `read` support for legacy format? 

how old is `legacy`? can we drop it? do we know, if `legacy` is even used?

Reply:
>we have only read support for legacy format?

How could you support writing a legacy format? It's always reading previous format versions and read-writing the current format version.

Reply:
can we drop the support for "legacy"? I suppose that we cannot, as we cannot know, if there is some stale journal, which hasn't been upgraded, right?

Reply:
>how old is legacy? can we drop it? do we know, if legacy is even used?

No idea, I kept the same logic of the test as before. I don't think I want to do soul-searching here - most likely old format are with us forever and ever.